### PR TITLE
wsd: fileserver.cpp changed variable replacemenct method

### PR DESCRIPTION
**Pull request by NoahbJohnson15**
Change-Id: Ie1a6c00f23a898c4244025f83d78cdd19ff73b9b

* Resolves: # 
* Target version: master 

### TODO

- [ ] ...

### Checklist

- [x] Code is properly formatted
- [x] All commits have Change-Id
- [x] I have run tests with `make check`
- [x] I have issued `make run` and manually verified that everything looks okay
- [x] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by pedropintosilva** [comment]
Hello @NoahbJohnson15 and welcome!

If I'm not mistaken this is related to https://github.com/CollaboraOnline/online/issues/4454 right? Let me assign @Ashod  as reviewer

**Review by pedropintosilva — CHANGES_REQUESTED** [review]
Could you please amend the commit and insert a commit subject (title) and commit message (description)?

* `git commit --ammend`
* And add 3 lines on top of `Signed-off-by: NoahJohnson <noahbjohnson18@gmail.com>`
  * `wsd: fileserver.cpp changed variable replacement method`
  * Empty line
  * `This speeds-up variable substitution #4454`

note: I also see a new added log file, I wonder if that's ok

File: browser/mocha.log [hunk 1]
```diff
@@ -0,0 +1,173 @@
+
+
+  Bounds parse() tests
+    Bounds.parse() call with an empty string argument
+incomplete rectangle
+      ✓ should return undefined
+    Bounds.parse() call with an string argument with 3 numbers
+incomplete rectangle
+      ✓ should return undefined
```
Comment (Ashod) [inline browser/mocha.log:9]: Since this is unrelated to the fileserver, best to leave it out. If need help with removing this file from your commit, just ask. Happy to help with git commands.

File: wsd/FileServer.cpp [hunk 2]
```diff
@@ -562,7 +562,8 @@ void FileServerRequestHandler::handleRequest(const HTTPRequest& request,
             endPoint == "uno-localizations.json" ||
             endPoint == "uno-localizations-override.json")
         {
-            preprocessFile(request, requestDetails, message, socket);
+            std::map<std::string, std::string> replaceVar = {{"ACCESS_TOKEN",""}, {"SOCKET_PROXY", ""}, {"ACCESS_TOKEN_TTL", ""}, {"ACCESS_HEADER", ""}, {"HOST", ""},  {"VERSION", "COOLWSD_VERSION_HASH"}, {"COOLWSD_VERSION", "COOLWSD_VERSION"}, {"SERVICE_ROOT", ""}, {"UI_DEFAULTS", ""}, {"POSTMESSAGE_ORIGIN", ""}, {"PROTOCOL_DEBUG", ""}, {"HEXIFY_URL", ""}, {"BRANDING_CSS", ""}, {"BRANDING_JS", ""},  {"CSS_VARIABLES", ""}, {"DOCUMENT_SIGNING_DIV", ""} , {"DOCUMENT_SIGNING_URL", ""}, {"BROWSER_LOGGING", ""}, {"GROUP_DOWNLOAD_AS", ""},  {"OUT_OF_TIMEOUT_SECS", ""}, {"IDLE_TIMEOUT_SECS", ""}, {"ENABLE_WELCOME_MSG", ""}, {"AUTO_SHOW_WELCOME", ""}, {"UI_RTL_SETTINGS", ""},   {"USE_INTEGRATION_THEME", ""}, {"ENABLE_MACROS_EXECUTION", ""},  {"AUTO_SHOW_FEEDBACK", ""}, {"FEEDBACK_URL", "FEEDBACK_URL"}, {"WELCOME_URL", "WELCOME_URL"}, {"BUYPRODUCT_URL", ""}, {"FRAME_ANCESTORS", ""}, {"USER_INTERFACE_MODE",""}};
```
Comment (Ashod) [inline wsd/FileServer.cpp:565]: Since we are setting the map keys and values below, I wonder what advantage is there in having default values that we later override. It seems that we only need default values for the cases that we don't have a value for. So we can simplify this greatly.

This line is excessively long. We use clang-format, which you can also use to automatically format all your code, but you can also manually break this on multiple lines, roughly making each line equal in length (about 80-100 characters long).

This helps readability and maintainability.

File: wsd/FileServer.cpp [hunk 3]
```diff
@@ -872,7 +873,8 @@ namespace
 void FileServerRequestHandler::preprocessFile(const HTTPRequest& request,
                                               const RequestDetails &requestDetails,
                                               Poco::MemoryInputStream& message,
-                                              const std::shared_ptr<StreamSocket>& socket)
+                                              const std::shared_ptr<StreamSocket>& socket,
+                                              std::map<std::string, std::string> replaceVar)
```
Comment (Ashod) [inline wsd/FileServer.cpp:877]: Passing the argument as reference (`std::map<std::string, std::string>& replaceVar`) is better.

File: wsd/FileServer.cpp [hunk 4]
```diff
@@ -940,29 +942,27 @@ void FileServerRequestHandler::preprocessFile(const HTTPRequest& request,
     std::string socketProxy = "false";
     if (requestDetails.isProxy())
         socketProxy = "true";
-    Poco::replaceInPlace(preprocess, std::string("%SOCKET_PROXY%"), socketProxy);
+    replaceVar["SOCKET_PROXY"] = socketProxy;
```
Comment (Ashod) [inline wsd/FileServer.cpp:945]: Shouldn't we include the `%` too? Why do we drop the percentage marks from all the variables?

File: wsd/FileServer.cpp [hunk 5]
```diff
@@ -984,9 +984,9 @@ void FileServerRequestHandler::preprocessFile(const HTTPRequest& request,
     }
 #endif
 
-    Poco::replaceInPlace(preprocess, std::string("<!--%BRANDING_CSS%-->"), brandCSS);
-    Poco::replaceInPlace(preprocess, std::string("<!--%BRANDING_JS%-->"), brandJS);
-    Poco::replaceInPlace(preprocess, std::string("<!--%CSS_VARIABLES%-->"), cssVarsToStyle(cssVars));
+    replaceVar["BRANDING_CSS"] = brandCSS;
+    replaceVar["BRANDING_JS"] = brandJS;
+    replaceVar["CSS_VARIABLES"] = cssVarsToStyle(cssVars);
```
Comment (Ashod) [inline wsd/FileServer.cpp:989]: Shouldn't we retain the "" as well? We then wouldn't need to special case in the replace loop below.

File: wsd/FileServer.cpp [hunk 6]
```diff
@@ -1116,14 +1114,30 @@ void FileServerRequestHandler::preprocessFile(const HTTPRequest& request,
...
         LOG_TRC("Denied all frame ancestors");
         cspOss << imgSrc << "; ";
     }
 
+    std::string newpreprocess;
+    newpreprocess.reserve(preprocess.length() + 50);
+    std::stringstream checkVar(preprocess);
+    std::string medium;
+
+     while(getline(checkVar, medium, '$'))
```
Comment (Ashod) [inline wsd/FileServer.cpp:1130]: Isn't the variable marker `%` ? Perhaps a comment to explain what this loop is doing is in order.

File: wsd/FileServer.cpp [hunk 7]
```diff
@@ -1116,14 +1114,30 @@ void FileServerRequestHandler::preprocessFile(const HTTPRequest& request,
...
+     {
+        if(medium.substr(medium.length()-4, medium.length()) == "<!--")
+            medium = medium.substr(0, medium.length()-4);
+        if(medium.substr(0, 3) == "-->")
+            medium = medium.substr(3, medium.length());
```
Comment (Ashod) [inline wsd/FileServer.cpp:1135]: Why do we delete the comment markers? Was that done in the old algorithm?

File: wsd/FileServer.cpp [hunk 8]
```diff
@@ -1116,14 +1114,30 @@ void FileServerRequestHandler::preprocessFile(const HTTPRequest& request,
...
     }
 
+    std::string newpreprocess;
+    newpreprocess.reserve(preprocess.length() + 50);
+    std::stringstream checkVar(preprocess);
+    std::string medium;
+
+     while(getline(checkVar, medium, '$'))
+     {
+        if(medium.substr(medium.length()-4, medium.length()) == "<!--")
```
Comment (Ashod) [inline wsd/FileServer.cpp:1132]: `substr` is rather expensive, because it creates a copy. We can improve this (perhaps later).

Also, best to follow the formatting of the file, for consistency. Leave space after the `while` and `if`, and add a line break after `else`. You can also use clang-format on the lines you've changed/added.

**Review by Ashod — CHANGES_REQUESTED** [review]
Thanks @NoahbJohnson15 for the patch! 

I left some comments and questions. The formatting and performance issues are minor. The ones related to the functionality are more critical, of course.

In addition, I'm curious how you tested this. It would be best to diff the output of this function before and after your change. They should be identical. Also, it's a good idea to add new test to verify your algorithm.

Other than that, there are test failures. But we'll get to that after you update the patch with the requested changes and we'll see if it still fails or not. So don't worry about that just yet.

Finally, you can use the title from the issue you are fixing in the commit message, which makes things easier for everyone. For example:
```
wsd: speed-up variable substitution in file-server
```

Btw, even though your approach is not what is proposed in #4454, it's fine, as long as it works as expected. We can improve it further later. (There is a detailed description of the algorithm in one of my comments).

Thanks, and feel free to ask any questions you may have!

**Comment by pedropintosilva** [comment]
Hello @NoahbJohnson15 are you still working on this ? Or is there anything I can help with?
